### DISCUSSION TITLE
The main dependencies of MEEP calculations.

### DISCUSSION THREAD

**Discussion by mariusCZ** [NanoComp/meep#2414] [Q&A] [answered]
Hi everyone. I am using two HPCs, one has Intel Xeon Platinum 8358 (faster CPUs) and the other has Intel Xeon Processor E5-2640 v4 (slower CPUs). Both HPCs use the same SLURM scheduling system and on both systems the simulations were run with 64 cores. Interestingly enough, the system with faster CPUs ran the simulations 15% slower. However, it performed significantly faster on epsilon setting and subpixel smoothing.

My question is not how to solve this problem, but rather what are the dependencies that are used for the actual FDTD calculations? Because the first possible cause that seems plausible is the dependencies being compiled differently. My simulation code is as follows:

```
import meep as mp
import matplotlib.pyplot as plt
import numpy as np

resolution = 20

ind = 3.4
a = 240
h = 170/a
r = 70/240
r_l3 = r*(5/7)
r_w1 = 60/a
r_a = 74/a
r_b = 76/a
r_c = 76/a
r_d = 68/a
r_e = 76/a
wvg_w = 280/a
l3_shift = 0.1

shifts = np.array([0.2, 0.00254641, 0.2, 0.03472173, -0.13992794, 0.07866315, -0.0470081, 0.0578609, 0.07736117, 0.12185224, -0.01850456, 0.10533734])
shiftN = int(shifts.size/2)

fcen = a/900
df = 0.05

N = 30
pad = 4
dpml = 2

sx = pad+dpml+N
sy = sx * (np.sqrt(3)/2)
sz = 6

cell = mp.Vector3(sx, sy, sz)

blk = mp.Block(size=mp.Vector3(mp.inf, mp.inf, h), material=mp.Medium(index=ind))
geometry = [blk]

posh = np.zeros([N, N, 2])
rh = np.zeros([N, N])

# 0 is -N/2
for i in range(N):
    for j in range(N):
        if i % 2 != 0:
            posh[i, j, 0] = -N/2 + j
        else:
            posh[i, j, 0] = -N/2 + j + 0.5
        posh[i, j, 1] = (-N/2 + i) * (np.sqrt(3)/2)
        rh[i, j] = r

# Cavity radius setting
rh[int(N/2), int(N/2)] = 0
rh[int(N/2)+1, int(N/2)] = 0
rh[int(N/2)-1, int(N/2)-1] = 0
rh[int(N/2)+2, int(N/2)+1] = r_l3
rh[int(N/2)-2, int(N/2)-1] = r_l3

def rotate_shift(dx, dy):
    xsh = 0.5*dx - (np.sqrt(3)/2)*dy
    ysh = (np.sqrt(3)/2)*dx + 0.5*dy
    return [xsh, ysh]

#Cavity position setting
for i in range(2, shiftN+2):
    rot_shifts = rotate_shift(shifts[i-2], shifts[(i-2)+shiftN]*np.sqrt(3))
    posh[int(N/2)+i, int(N/2)+int(i/2), 0] += rot_shifts[0]
    posh[int(N/2)+i, int(N/2)+int(i/2), 1] += rot_shifts[1]

    rot_shifts = rotate_shift(-shifts[i-2], shifts[(i-2)+shiftN]*np.sqrt(3))
    posh[int(N/2)-i, int(N/2)-int((i+1)/2), 0] += rot_shifts[0]
    posh[int(N/2)-i, int(N/2)-int((i+1)/2), 1] += rot_shifts[1]
    
# Waveguide
rh[int(N/2)+2, int(N/2)+3:] = 0
rh[int(N/2)+3, int(N/2)+4:] = r_w1
rh[int(N/2)+1, int(N/2)+4:] = r_w1

# Perturbations
# A
# rh[int(N/2)+2, int(N/2)+2] = r_a
# B
# rh[int(N/2)+3, int(N/2)+2] = r_b
# C 
# rh[int(N/2)+3, int(N/2)+3] = r_c
# D
# rh[int(N/2)+1, int(N/2)+2] = r_d
# E
# rh[int(N/2)+1, int(N/2)+3] = r_e

for i in range(N):
    for j in range(N):
        if rh[i,j] != 0:
            geometry.append(mp.Cylinder(rh[i,j], center=mp.Vector3(posh[i, j, 0], posh[i, j, 1], 0), height = mp.inf))

pml_layers = [mp.PML(dpml, mp.X), mp.PML(dpml, mp.Y), mp.PML(1, mp.Z)]

src = [mp.Source(mp.GaussianSource(fcen, fwidth=df), mp.Ey, mp.Vector3(0, 0))]

sim = mp.Simulation(cell_size=cell,
                    boundary_layers=pml_layers,
                    geometry=geometry,
                    sources=src,
                    resolution=resolution)

freq = mp.FluxRegion(center=mp.Vector3(0.5*sx-dpml-6, np.sqrt(3)),
                     size=mp.Vector3(0,2*np.sqrt(3),2*h))

nfreq = 1000 # number of frequencies at which to compute flux

# transmitted flux
trans = sim.add_flux(fcen, df, nfreq, freq)

sim.plot2D(output_plane=mp.Volume(size=mp.Vector3(sx,sy)))
plt.savefig('sim_dom.png')

pt = mp.Vector3(0.5*sx-dpml-6, np.sqrt(3))

# sim.run(until_after_sources=mp.stop_when_fields_decayed(50,mp.Ey,pt,1e-3))
hinv = mp.Harminv(mp.Ey, mp.Vector3(), fcen, df)

sim.run(mp.after_sources(hinv),
        until_after_sources=400)
      
Qfac = []
amp = []  
for modes in hinv.modes:
	if modes.freq > 0.254 and modes.freq < 0.258:
		Qfac.append(modes.Q)
		amp.append(modes.amp)
		
print(Qfac[amp.index(max(amp))])
        
sim.display_fluxes(trans)

```

**Comment by stevengj** [comment, marked as answer]
Most of the time is spent in Meep itself, not in external libraries.

The speed of the memory is probably more important than the peak flop rate of the CPU, and it could well be that your "faster CPU" system has a slower memory hierarchy.

**Comment by mariusCZ** [comment]
Now that you mention it, I noticed some interesting behaviour. On the
system with "slower CPUs", 16 GB of memory was more than enough for the
simulation. However, on the other system, with 16 GB the simulation runs,
however, right at the end it crashes with an out of memory error. This is
fixed by increasing the memory. Could this be related perhaps?

**Reply by smartalecH** [reply]
This seems a bit odd to me... if you are running on the same number of cores between experiments (such that the number of chunks etc is the same) then the memory profile should be relatively identical too...

Are you sure the first system didn't just check out more memory without telling you?

Is this 16Gb _per process_? Or total? Different systems handle this constraint differently (even within the same slurm schedule profile).

So this could indicate something is wrong with your setup. 

But, like Steven said, your problem is probably due to communication overhead and fetching. You should look up the memory bandwidth of the two systems as a first-order check.

**Comment by oskooi** [comment]
See also: https://meep.readthedocs.io/en/latest/Parallel_Meep/#runtime-scaling-on-mpi-clusters.

**Comment by mariusCZ** [comment]
I'll check out the actual set memory, as it is quite a weird behaviour.

As for communication overhead, the newer system should be better, since one
node is 64 cores, whereas the older one is 4 nodes for 64 cores, which
should make it slower

**Reply by stevengj** [reply]
Not necessarily; with one node being 64 cores you could have more memory contention between the cores.
